tsl/src/compression/api.c

```diff
@@ -892,3 +896,584 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+												  Relation uncompressed_chunk_rel,
+												  CompressedSegmentInfo **current_segment)
+{
+	TableScanDesc heapScan;
+	HeapTuple uncompressed_tuple;
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+	ScanKeyData *scankey = palloc0(sizeof(*scankey) * nsegmentby_cols);
+
+	for (int seg_col = 0; seg_col < nsegmentby_cols; seg_col++)
+	{
+		Datum val = current_segment[seg_col]->segment_info->val;
+		// what is the attno of this segby col in the uncompressed chunk?
+		int16 attno =
+			current_segment[seg_col]->decompressed_chunk_offset + 1; // offset is attno - 1
+		ScanKeyEntryInitializeWithInfo(&scankey[seg_col],
+									   0 /*flags*/,
+									   attno,
+									   BTEqualStrategyNumber,
+									   InvalidOid,
+									   current_segment[seg_col]->segment_info->collation,
+									   &current_segment[seg_col]->segment_info->eq_fn,
+									   val);
+	}
+
+	heapScan =
+		table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), nsegmentby_cols, scankey);
+
+	for (uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+		 uncompressed_tuple != NULL;
+		 uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	{
+		TupleTableSlot *heap_tuple_slot =
+			MakeTupleTableSlot(uncompressed_rel_tupdesc, &TTSOpsHeapTuple);
+
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(segment_tuplesortstate, heap_tuple_slot);
+		/* simple_heap_delete since we don't expect concurrent updates, have exclusive lock on the
+		 * relation */
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+		ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	}
+	table_endscan(heapScan);
+	pfree(scankey);
+}
+/*
+ * This function will replace the procedure for recompress_chunk
+ * It will keep the compressed chunk around, only decompressing the
+ * batches that are affected by the addition of newer data.
+ * If more batches need to be created, it will create those and update
+ * the meta_sequence_number of the following ones.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed =
+		PG_ARGISNULL(1) ? false : PG_GETARG_BOOL(1); // probably don't need this
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (NULL == uncompressed_chunk)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
+
+	/* need to push down to data nodes if this is an access node */
+	if (uncompressed_chunk->relkind == RELKIND_FOREIGN_TABLE)
+	{
+		if (!recompress_remote_chunk(fcinfo, uncompressed_chunk))
+			PG_RETURN_NULL();
+
+		/* TODO: probably redundant? */
+		ts_chunk_set_compressed_chunk(uncompressed_chunk, INVALID_CHUNK_ID);
```

Comment:
Could you check this in the final version of the PR?

Reply:
It was redundant because I am not creating a new compressed chunk, and for the existing compressed chunk, this is already marked Invalid upon its creation. So I removed it.